proposed/2023/vs-filters.md

```diff
@@ -0,0 +1,89 @@
+
+- [Jon Douglas](https://github.com/JonDouglas)
+- 2023-06-23
+
+## Summary
+
+<!-- One-paragraph description of the proposal. -->
+This proposal represents a common ask for more filters and sorting options in the NuGet Package Manager experience in Visual Studio. 
+
+## Motivation 
+
+<!-- Why are we doing this? What pain points does this solve? What is the expected outcome? -->
+As NuGet continues to expand as a package ecosystem, there has become a dire need for being able to filter and sort packages through one's daily management in their IDE of choice. While NuGet.org supports a plethora of filters and sorts on the search experience today, we'd like to bring these similar experiences to the various tabs(browse, installed, etc) in Visual Studio to help aid various jobs-to-be-done when it comes down to package management.
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or pseudocode to show how this proposal would look. -->
+
+When a developer is using Visual Studio to search and manage NuGet packages, they will see a new filter and sort experience. 
+
+The browse filter experience will allow the developer to filter results based on common concepts such as "frameworks", "prerelease", "owners", "download count", "license", and "created on". 
+
+![](resources/VSFilter/PackageOptions.png)
+
+The installed filter experience will allow the developer to filter results based on all the **browse filter** and additional package status filters such as "deprecated", "vulnerable", and "update".
+
+![](resources/VSFilter/PackageOptionsAndStatus.png)
+
+To apply filters, the developer will click the new filter button next to the search box. This will open up a new window of the possible filtering options to choose from.
+
+When a filter is selected, the filter will show up as a pill/badge affordance in the package list which can be removed by clicking on the "X" associated with the pill/badge. For some filters, they will support multiple values such as frameworks, owners, and licenses. For other filters, they will only support one value such as prerelease, download count, created on, etc. For filters that have popular values but also have custom values, a input box will be provided. For example the license filter will have popular OSS licenses, but with the ability to insert the name of a not well known license as well.
+
+![](resources/VSFilter/PillsAndSort.png)
+
+Whenever a filter is applied, it will additionally show up in the Visual Studio search box so that developers get used to the search prefix syntax and can apply filters more quickly in the future.
+
+The sorting experience will allow the developer to sort these options on common concepts such as "alphabetical", "version", "relevance", "downloads", and "recently updated".
```

Comment:
I'm guessing this is jsut examples, rather than a concrete proposal?

Reply:
these are concrete examples 🥲 can go into more details on them if needed.

Reply:
Ah, I see :D My feedback on the ones I have concerns with. 

- version - Can you elaborate more? I'm not sure what this is supposed to mean or how is it helpful.
- relevance - can we define relevance? How would we measure it.
- downloads, recently updated - these are "network dependent", and what I mean by that is that it needs the client to reach out to the server and potentially change the order of what the customer is, so there's a good chance you'd have a noticeable flicker. This could get weird. Personally, I'd steer clear on any sorting that cannot be done locally, but I know there's been asks. 

Another side question, are the filters sticky? Within same session?

Reply:
version would sort by the package version in asc/desc order.

relevance would be the default search experience today just as a filter to get back to (like most search experiences)

downloads / recently updated - could we not cache the package details and use that locally?

At the time of writing, filters are not sticky as typical filtering and sorting experiences are not sticky outside the context of the same session. In other words, if a user exits the PM UI, the defaults will be set again. Further customer feedback may change this behavior.

Reply:
another thought might be that version could possibly be helpful if you can specify a range to filter as well. that might be helpful for the .NET libraries case and upgrading those in bulk.

Reply:
I actually missed the fact that some of these filters are supposed to apply on the search tab as well. 

For example, "relevance" is probably only a thing on the search tab. 
Honestly, alphabetical search on the search tab is probably...not very useful :D 

> downloads / recently updated - could we not cache the package details and use that locally?

Downloads is non trivial. It's only available in search, so it'd be super costly to get that data for the installed table.
Recently might be doable since registration will be hit for those packages anyways.

Reply:
Sounds good. We'll have to outline feasible "Browse" filters/sorts and "Installed" filters/sorts then based on cost if we don't get them for free.

Recently updated would be any version. Similar to NuGet.org -> https://www.nuget.org/packages?q=&frameworks=&tfms=&packagetype=&prerel=true&sortby=created-desc

Reply:
I think recently updated is sort of doable, but it has the same "flicker" potential as the other installed tab sorts. 
Not too worried about the search tab as that'll come from the server.

I think we should call out that sorts like that won't work on all servers and potentially if "All" is selected.

Reply:
I think versions should allow package versions and assembly versions. The reason is that if there is an assembly not found exception or similar issue, the error returns the actual assembly version rather than the package version. Right now my process is to click Show All Files in Solution Explorer, search for the problematic dll by name, and then look at the Dependencies tree to find the concrete dll version used by clicking into it.

Reply:
Right now our concept would only support package metadata which would be a package version. If an actual assembly version does not match the package version, that may be better resolved with the package author.

Reply:
> Right now our concept would only support package metadata which would be a package version. If an actual assembly version does not match the package version, that may be better resolved with the package author.

That is impractical in the case when an author ships a hotfix for an older framework. In any case, completely disagree as standard practice among many top 100 non Microsoft packages is major.0.0.0 precisely because this is a PITA to deal with. Microsoft however refuses this convention for OOB packages.

Please consider this filter.

Reply:
Understood.  I think some people _don't know to ask for it_, though. It's like the Steve Jobs principle of listening to your customers (we need better searching), but then _figuring out what they should really have help searching for, rather than just building the ones they ask for_.  When you consider that .NET Tree Shaking of assemblies removes classes/methods, the possibility of version conflicts grows and the ultimate debugging details lie in what assembly you are using, not a nuget package version. - When .NET throws an assembly-related exception, the error message is always from the point of view of the assembly, not the package.

I have a pretty convoluted process to work around these problems, personally, and I've mastered it, so it doesn't really affect me any more since I have now become an expert at it, but boy, do I notice _a lot_ of people struggle with this. And the messaging that .NET Core took care of this stuff with deps.json generation is just wrong - netstandard2.0 has an unbounded upper cardinality on what .NET 5+ version it could accept, etc.

Anyway, will table this aspect of the review for now - assuming there is a separate later phase documented so this doesnt get lost.